components/support/sync-telemetry/src/main/java/mozilla/components/support/sync/telemetry/SyncTelemetry.kt

```diff
@@ -4,28 +4,78 @@
 
 package mozilla.components.support.sync.telemetry
 
+import mozilla.appservices.sync15.EngineInfo
 import mozilla.appservices.sync15.FailureName
 import mozilla.appservices.sync15.FailureReason
 import mozilla.appservices.sync15.SyncTelemetryPing
 import mozilla.components.service.glean.private.LabeledMetricType
 import mozilla.components.service.glean.private.StringMetricType
+import mozilla.components.support.base.log.logger.Logger
 import mozilla.components.support.sync.telemetry.GleanMetrics.BookmarksSync
 import mozilla.components.support.sync.telemetry.GleanMetrics.HistorySync
 import mozilla.components.support.sync.telemetry.GleanMetrics.LoginsSync
+import mozilla.components.support.sync.telemetry.GleanMetrics.Sync
 import mozilla.components.support.sync.telemetry.GleanMetrics.Pings
+import java.util.UUID
 
 const val MAX_FAILURE_REASON_LENGTH = 100
 
 /**
  * Contains functionality necessary to process instances of [SyncTelemetryPing].
  */
+@Suppress("LargeClass")
 object SyncTelemetry {
+    private val logger = Logger("SyncTelemetry")
+
+    /**
+     * Process [SyncTelemetryPing] as returned from [mozilla.appservices.syncmanager.SyncManager].
+     */
+    fun processSyncTelemetry(syncTelemetry: SyncTelemetryPing) {
+        syncTelemetry.syncs.forEach { syncInfo ->
+            val syncUuid = UUID.randomUUID()
+            Sync.syncUuid.set(syncUuid)
+
+            // It's possible for us to sync some engines, and then get a hard error that fails the
+            // entire sync. Examples of such errors are an HTTP server error, token authentication
+            // error, or other kind of network error.
+            // We can have some engines that succeed (and others that fail, with different reasons)
+            // and still have a global failure_reason.
+            syncInfo.failureReason?.let { failureReason ->
+                recordFailureReason(failureReason, Sync.failureReason)
+            }
+
+            syncInfo.engines.forEach { engineInfo ->
+                when (engineInfo.name) {
+                    "bookmarks" -> {
+                        individualBookmarksSync(syncTelemetry.uid, syncUuid, engineInfo)
```

Comment:
If you change the metric definition as suggested by @linacambridge , then you will be able to set the uuid only once.